x-pack/plugins/lists/server/services/extension_points/extension_point_storage_client.ts

```diff
@@ -0,0 +1,95 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { Logger } from 'kibana/server';
+
+import { ExtensionPoint, NarrowExtensionPointToType } from './types';
+import { ExtensionPointError } from './errors';
+import { ExtensionPointStorageInterface } from './extension_point_storage';
+
+export class ExtensionPointStorageClient {
+  constructor(
+    private readonly storage: ExtensionPointStorageInterface,
+    private readonly logger?: Logger
+  ) {}
+
+  /**
+   * Retrieve a list (`Set`) of extension points that are registered for a given type
+   * @param extensionType
+   */
+  get<T extends ExtensionPoint['type']>(
+    extensionType: T
+  ): Set<NarrowExtensionPointToType<T>> | undefined {
+    return this.storage.get(extensionType);
+  }
+
+  /**
+   * Runs a set of callbacks by piping the Response from one extension point callback to the next callback
+   * and finally returning the last callback payload.
+   *
+   * @param extensionType
+   * @param initialCallbackInput The initial argument given to the first extension point callback
+   * @param callbackResponseValidator A function to validate the returned data from an extension point callback
+   */
+  async pipeRun<
+    T extends ExtensionPoint['type'],
+    D extends NarrowExtensionPointToType<T> = NarrowExtensionPointToType<T>,
+    P extends Parameters<D['callback']> = Parameters<D['callback']>
+  >(
+    extensionType: T,
+    initialCallbackInput: P[0],
+    callbackResponseValidator?: (data: P[0]) => Error | undefined
+  ): Promise<P[0]> {
+    let inputArgument: P[0] = initialCallbackInput;
+    const externalExtensions = this.get(extensionType);
+
+    if (!externalExtensions || externalExtensions.size === 0) {
+      return inputArgument;
+    }
+
+    for (const externalExtension of externalExtensions) {
+      const extensionRegistrationSource =
+        this.storage.getExtensionRegistrationSource(externalExtension);
+
+      try {
+        // FIXME:PT investigate if we can avoid the TS ignore below?
+        // @ts-expect-error
+        inputArgument = await externalExtension.callback(inputArgument);
```

Comment:
This should work it looks like:

```ts
        inputArgument = await externalExtension.callback(
          inputArgument as CreateExceptionListItemOptions & UpdateExceptionListItemOptions
        );
```

And you can remove the `@ts-expect-error` and that should get type safety back for the rest of the line.

Reply:
@FrankHassanabad ,

I know you suggested this cast in the test as well, but it will not scale as we add more callbacks. In my next PR, the arguments to these callbacks are going to change (we need to provide the `Request` and possibly `Context` to these external callbacks) and I will also be introducing new extension points.

Yesterday I actually attempted to get an intersection of all callback arguments, but failed at that. I now realize it was because a) I was not un-wrapping the promise returned by the callbacks and b) I was not converting the Union of return types to an intersection. So I just tried this, using existing TS utility types available in kibana modules, and seems to work:

```typescript
type AnyArg = UnionToIntersection<
  PromiseType<
    ReturnType<ExtensionPoint['callback']>
    >
>;
```

Unless you see something wrong with this approach, I will add this approach instead.